### ISSUE TITLE
some question about mmseg-related Dataloader

### ISSUE THREAD

**Issue by endaoguansanlu** [open-mmlab/mmcv#2377]
Hello mmcv developers! I used a loss module in the process of using mmsegmentation, but this module is very slow to load in a single GPU, so I want to load it in the Dataloader, but I don't know how to modify the mmseg-related Dataloader in mmcv. So I hope you developers can give me a little guidance, thank you!

Below is a demo of the module I want to use loaded in Dataloader.

import torch
from torch.utils.data import Dataset, DataLoader

```
import numpy as np
from scipy.ndimage import distance_transform_edt as distance
from torchvision import transforms
from functools import partial
from operator import itemgetter

from tqdm import tqdm
import cv2


def kl_div(a,b): 
    return F.softmax(b, dim=1) * (F.log_softmax(b, dim=1) - F.log_softmax(a, dim=1))   

def one_hot2dist(seg):
    res = np.zeros_like(seg)
    for i in range(len(seg)):
        posmask = seg[i].astype(np.bool)
        if posmask.any():
            negmask = ~posmask
            res[i] = distance(negmask) * negmask - (distance(posmask) - 1) * posmask
    return res

def class2one_hot(seg, C):
    seg = seg.unsqueeze(dim=0) if len(seg.shape) == 2 else seg
    res = torch.stack([seg == c for c in range(C)], dim=1).type(torch.int32)
    return res

class DummyDataset(Dataset):
        def __init__(self):
            self.dist_map_transform = transforms.Compose([
                lambda img: img.unsqueeze(0),
                lambda nd: nd.type(torch.int64),
                partial(class2one_hot, C=1),
                itemgetter(0),
                lambda t: t.cpu().numpy(),
                one_hot2dist,
                lambda nd: torch.tensor(nd, dtype=torch.float32)
            ])
        
        def get_dist_maps(self, target):
            target_detach = target.clone().detach()
            dist_maps = torch.cat([self.dist_map_transform(target_detach[i]) for i in range(target_detach.shape[0])])
            out = -dist_maps
            out = torch.where(out>0, out, torch.zeros_like(out))
            return out

        def __len__(self):
                return 100

        def __getitem__(self, index):
                dummy_gt_np = np.random.randint(-10,2,[5,20,20]) # random, onehot, [gt_class, h, w]
                dummy_gt_np[dummy_gt_np<0] = 0 # make it sparse
                dummy_gt = torch.from_numpy(dummy_gt_np)
                dist = self.get_dist_maps(dummy_gt)
                return dummy_gt_np, dist

if __name__ == '__main__':
    dummy_dataset = DummyDataset()
    dummy_loader = DataLoader(dummy_dataset, batch_size=2)
    for i, (gt, dist) in enumerate(tqdm(dummy_loader)):
        # cv2.imwrite(f'vis/{i}.png', gt[0][0].numpy().astype(np.uint8)*255)
        # cv2.imwrite(f'vis/{i}_dist.png', dist[0][0].numpy().astype(np.uint8) * 10)
        pass
```

**Comment by HAOCHENYE** [comment]
Hi! thanks for your feedback. MMSeg has implemented [CE-loss](https://github.com/open-mmlab/mmsegmentation/blob/eb5ea47d9cb8cee98cdf2d5b37c027b6a1206ef7/mmseg/models/losses/cross_entropy_loss.py#L12) and you can use it directly! 
BTW, why we need to load the loss module in `dataset`? (or a `Dataloader`? The demo code seems only customize a dataset, not `Dataloader`).

**Comment by endaoguansanlu** [comment]
Thanks for your quick reply! I want to use CELoss with this Loss for multiple loss functions. The `def get_dist_maps(self, target)`  in this loss runs very slow in GPU, the inventor of this loss told me that i can speed it up by moving the process of distance map generation(`def get_dist_maps(self, target)`) to the Dataloader, in this way, maybe I need to rewrite some code of mmcv. So I have this question  T_T

**Comment by HAOCHENYE** [comment]
Alright, it seems the core requirements is how to move  `get_dist_maps` to dataset, rather than `DataLoader`. Since Dataloader will spawn multiple processes to accelerate fetching data from dataset, so implementing `get_dist_maps` in the dataset will have an obvious acceleration.

You can implement a custom pipeline in https://github.com/open-mmlab/mmsegmentation/tree/master/mmseg/datasets/pipelines, and than `get_dist_maps` will be called in dataset.

**Comment by endaoguansanlu** [comment]
Thank you so much for your patient advice! ! ! I will try to implement it! !
